### PR TITLE
[llvm-objcopy] --gap-fill and 0-size sections

### PR DESCRIPTION
**Pull request by quic-akaryaki**
In the change that added `--gap-fill`, the condition to choose the sections to write in `BinaryWriter::write()` did not exclude zero-size sections. However, zero-size sections did not have correct offsets assigned in `BinaryWriter::finalize()`. The result is either a failed assertion, or memory corruption due to writing to the buffer beyond its size.
To fix this, exclude zero-size sections and add a zero-size section to the test, which would trigger the bug.

### CODE REVIEW

**Review by jh7370 — APPROVED** [review]
LGTM, with nits.

File: llvm/test/tools/llvm-objcopy/ELF/gap-fill.test [hunk 1]
```diff
@@ -106,6 +106,11 @@ Sections:
     Flags:           [ SHF_ALLOC, SHF_EXECINSTR ]
     Address:         0x0108
     Content:         'AABBCCDDFEDCBA'
+  - Name:            .zero_size
+    Type:            SHT_PROGBITS
+    Flags:           [ SHF_ALLOC, SHF_EXECINSTR ]
+    Address:         0x0110
+    Size:            0
```
[the remark this answers is not in the document]
Comment (quic-akaryaki) [reply llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:113]: Yes, I'd like to leave it.

File: llvm/test/tools/llvm-objcopy/ELF/gap-fill.test [hunk 2]
```diff
@@ -106,6 +106,11 @@ Sections:
     Flags:           [ SHF_ALLOC, SHF_EXECINSTR ]
     Address:         0x0108
     Content:         'AABBCCDDFEDCBA'
+  - Name:            .zero_size
+    Type:            SHT_PROGBITS
+    Flags:           [ SHF_ALLOC, SHF_EXECINSTR ]
```
[the remark this answers is not in the document]
Comment (quic-akaryaki) [reply llvm/test/tools/llvm-objcopy/ELF/gap-fill.test]: Done

**Comment by quic-akaryaki** [comment]
Thank you for the feedback @jh7370 .

**Pull request merged**